File: Day030_2025/Q030_GFG.java
import java.util.*;

class Q030_GFG {
    private ArrayList<ArrayList<Integer>> result = new ArrayList<>();
    private int[] row;

    private boolean isSafe(int r, int c) {
        for (int prev = 0; prev < c; prev++) {
            if (row[prev] == r || Math.abs(row[prev] - r) == Math.abs(prev - c)) {
                return false;
            }
        }
        return true;
    }

    private void solveNQueens(int column, int n) {
        if (n == 2 || n == 3) return;
        
        if (column == n) {
            ArrayList<Integer> temp = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                temp.add(row[i] + 1);
            }
            result.add(temp);
            return;
        }

        for (int i = 0; i < n; i++) {
            if (isSafe(i, column)) {
                row[column] = i;
                solveNQueens(column + 1, n);
            }
        }
    }

    public ArrayList<ArrayList<Integer>> nQueen(int n) {
        result.clear();
        row = new int[n];
        solveNQueens(0, n);
        return result;
    }
}

File: Day030_2025/Q030_GFG.py
class Solution:
    def __init__(self):
        self.result = []
        self.row = []

    def isSafe(self, r, c):
        for prev in range(c):
            if self.row[prev] == r or abs(self.row[prev] - r) == abs(prev - c):
                return False
        return True

    def solveNQueens(self, column, n):
        if n == 2 or n == 3:
            return
        
        if column == n:
            self.result.append([self.row[i] + 1 for i in range(n)])
            return
        
        for i in range(n):
            if self.isSafe(i, column):
                self.row[column] = i
                self.solveNQueens(column + 1, n)

    def nQueen(self, n):
        self.result = []
        self.row = [-1] * n
        self.solveNQueens(0, n)
        return self.result
